src/backend/EdFi.DataManagementService.Backend.Postgresql/Operation/SqlAction.cs

```diff
@@ -481,26 +528,97 @@ NpgsqlTransaction transaction
     }
 
     /// <summary>
-    /// Delete a document for a given Id and returns the number of rows affected
+    /// Insert a set of rows into the References table and return the number of rows affected
     /// </summary>
-    public async Task<int> DeleteDocumentByDocumentId(
+    public async Task<int> InsertReferences(
+        BulkReferences bulkReferences,
+        NpgsqlConnection connection,
+        NpgsqlTransaction transaction
+    )
+    {
+        Trace.Assert(
+            bulkReferences.ReferentialIds.Length == bulkReferences.ReferentialPartitionKeys.Length,
+            "Arrays of ReferentialIds and ReferentialPartitionKeys must be the same length"
+        );
+
+        long[] parentDocumentIds = new long[bulkReferences.ReferentialIds.Length];
+        Array.Fill(parentDocumentIds, bulkReferences.ParentDocumentId);
+
+        int[] parentDocumentPartitionKeys = new int[bulkReferences.ReferentialIds.Length];
+        Array.Fill(parentDocumentPartitionKeys, bulkReferences.ParentDocumentPartitionKey);
+
+        await using var insertBulkReferencesCmd = new NpgsqlCommand(
+            @"INSERT INTO public.""references""(ParentDocumentId, ParentDocumentPartitionKey, ReferentialId, ReferentialPartitionKey)
+                    SELECT * FROM unnest($1, $2, $3, $4)",
```

Comment:
The `unnest()` magic. Provide arrays for each column value, and this SQL turns them into an insert across the columns for each array index value.

Reply:
very nice